### PR TITLE
EVG-12433 consider exec tasks when scheduling display task

### PR DESCRIPTION
**Pull request by ablack12**
We were only scheduling the display task, which means the exec tasks couldn't get scheduled in any way.

Whether or not this fix is temporary depends on what we do with display/execs in the future but 🤷  the fix is done so.

### CODE REVIEW

File: model/version.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ type Version struct {
 	GitTags           []GitTag `bson:"git_tags,omitempty" json:"git_tags,omitempty"`
 	TriggeredByGitTag GitTag   `bson:"triggered_by_git_tag,omitempty" json:"triggered_by_git_tag,omitempty"`
 
-	// Parameters stores both user-defined parameters and default parameters
+	// Parameters stores user-defined parameters
```
Comment (ablack12) [inline model/version.go:44]: Sorry for the unrelated comment change, I realized it was confusing when looking at something else

File: model/task_lifecycle.go [hunk 2]
```diff
@@ -32,24 +32,33 @@ type StatusChanges struct {
...
-			deps, err := task.GetRecursiveDependenciesUp([]task.Task{*t}, nil)
+			deps, err := task.GetRecursiveDependenciesUp(originalTasks, nil)
 			if err != nil {
 				return errors.Wrapf(err, "error getting tasks '%s' depends on", t.Id)
 			}
 			tasksToActivate = append(tasksToActivate, deps...)
 		}
 
-		if t.DispatchTime != utility.ZeroTime && t.Status == evergreen.TaskUndispatched {
+		if !utility.IsZeroTime(t.DispatchTime) && t.Status == evergreen.TaskUndispatched {
```
Comment (ybrill) [inline model/task_lifecycle.go:56]: maybe we need to do this check for every task in `originalTasks`?
I don't really understand why we do this check to begin with. Will it ever be true?

Comment (ablack12) [reply model/task_lifecycle.go:56]: This is a very good question. I'm having trouble figuring out where it was added; evidently we've had this line for years (i.e. it was taken from another file 5 years ago, where the line was the same). I wonder if it should've been `not the zero time && not undispatched`? Because why would you want to reset an undispatched task? I'm not sure if activeState is ever even called for that case though. So maybe the answer is to just get rid of this if statement?

We wouldn't need to check this for execution tasks, because the status of the display task reflects the progress of all the execution tasks, and `resetTask` can't be called on execution tasks.

Comment (ablack12) [reply model/task_lifecycle.go:56]: Actually I'm just kind of nervous to do this since I don't understand it. What if I add a log here to see if it gets hit? I can check in on it in a bit.

Comment (ybrill) [reply model/task_lifecycle.go:56]: I went on an expedition in old code (89a9143). It seems when a task was aborted the [agent would send back an undispatched status](https://github.com/evergreen-ci/evergreen/blob/89a914392443f9aadbdae327d289430e728602d8/agent/agent.go#L266) in EndTask. If you try to activate such a task (with an undispatched status but it's really been dispatched, by virtue of its dispatch time) what we really need to do is reset the task.

This combination was also part of task's [IsFinished](https://github.com/evergreen-ci/evergreen/blob/89a914392443f9aadbdae327d289430e728602d8/model/task.go#L228) until [Richard removed it](https://github.com/evergreen-ci/evergreen/commit/fca22796c7e32fbae6d6df4f7dc4d89183398030).

Given [this state no longer exists](https://github.com/evergreen-ci/evergreen/commit/ebf6771fae3f87e6f859fe300ca63d5b903209d5#diff-15bde6705d6dca44ef9236255f41880fL29), it's probably safe to remove.
I could be totally off on this, though...

Comment (ybrill) [reply model/task_lifecycle.go:56]: > We wouldn't need to check this for execution tasks, because the status of the display task reflects the progress of all the execution tasks, and resetTask can't be called on execution tasks.

Would we need to call [checkResetDisplayTask](https://github.com/evergreen-ci/evergreen/blob/a1d2c8f70bf5c543de8b9641ac1ec08def1ddb26/model/task_lifecycle.go#L1286) then?

Comment (ablack12) [reply model/task_lifecycle.go:56]: resetTask() handles execution tasks. I'm in favor of removing this though.

Comment (ablack12) [reply model/task_lifecycle.go]: Actually so, I ran this aggregation:
```
db.tasks.aggregate([
    {"$match" : {
		"status" : "undispatched",
		"dispatch_time" : {
			"$nin" : [
				ISODate("1970-01-01T00:00:00Z"),
				ISODate("0001-01-01T00:00:00Z")
			]}}},
    {"$project": {"dispatch_time":1, "_id":0}},
    {"$sort": {"dispatch_time":-1}},
    {"$limit":5}
])
```

and I did get 2020 results: 
```
{ "dispatch_time" : ISODate("2020-10-01T17:25:08.820Z") }
{ "dispatch_time" : ISODate("2020-10-01T17:24:05.660Z") }
{ "dispatch_time" : ISODate("2020-10-01T17:23:03.208Z") }
{ "dispatch_time" : ISODate("2020-10-01T17:20:46.170Z") }
```
I'm not sure what's going on here; I can make a ticket to investigate but I think let's consider it out of scope of this ticket?

Comment (ybrill) [reply model/task_lifecycle.go]: It looks to me like all those are display tasks in progress. I think the issue is that [MarkAsDispatched](https://github.com/evergreen-ci/evergreen/blob/a1d2c8f70bf5c543de8b9641ac1ec08def1ddb26/model/task/task.go#L658) (called from NextTask) is supposed to set display tasks as dispatched as soon as an execution task starts running. It doesn't, though, because the task document comes from [assignNextAvailableTask](https://github.com/evergreen-ci/evergreen/blob/a1d2c8f70bf5c543de8b9641ac1ec08def1ddb26/service/api_task.go#L445) which doesn't fetch the display task so we don't know the task is an execution task...

It's okay with me to leave this for another ticket if that works better. I don't understand what you meant though:
>resetTask() handles execution tasks

Didn't you say
> resetTask can't be called on execution tasks

Comment (ablack12) [reply model/task_lifecycle.go]: sorry, resetTask() calls MarkOneTaskReset which also marks all execution tasks. https://github.com/evergreen-ci/evergreen/blob/a5f3546bd358ff228c4ea707264536be14ea7c1e/model/task_lifecycle.go#L921

Comment (ybrill) [reply model/task_lifecycle.go:56]: Ah, I  understand now. But I had understood what you were saying that resetTask [errors on execution tasks](https://github.com/evergreen-ci/evergreen/blob/a1d2c8f70bf5c543de8b9641ac1ec08def1ddb26/model/task_lifecycle.go#L131).

**Review by ybrill — APPROVED** [review]

File: model/task_lifecycle.go (hunk 2 above)
Comment (ablack12) [reply model/task_lifecycle.go:56]: Sorry poorly worded on my part 😅  Here is a ticket to investigate the wonky display task state. 
https://jira.mongodb.org/browse/EVG-13144
